File: X-AI/LIME/C++/lime.cpp
#include <bits/stdc++.h>
using namespace std;
vector<double> linear_regression(const vector<vector<double>> &X, const vector<double> &y, const vector<double> &w)
{
    int d = X[0].size();
    vector<vector<double>> A(d, vector<double>(d, 0));
    vector<double> b(d, 0);
    for (size_t i = 0; i < X.size(); i++)
    {
        double wi = w[i];
        for (int j = 0; j < d; j++)
        {
            b[j] += wi * y[i] * X[i][j];
            for (int k = 0; k < d; k++)
                A[j][k] += wi * X[i][j] * X[i][k];
        }
    }
    for (int j = 0; j < d; j++)
        A[j].push_back(b[j]);
    for (int i = 0; i < d; i++)
    {
        int piv = i;
        for (int r = i + 1; r < d; r++)
            if (fabs(A[r][i]) > fabs(A[piv][i]))
                piv = r;
        swap(A[i], A[piv]);
        if (fabs(A[i][i]) < 1e-12)
            continue;
        double fac = A[i][i];
        for (int c = i; c <= d; c++)
            A[i][c] /= fac;
        for (int r = 0; r < d; r++)
            if (r != i)
            {
                double f = A[r][i];
                for (int c = i; c <= d; c++)
                    A[r][c] -= f * A[i][c];
            }
    }
    vector<double> beta(d);
    for (int j = 0; j < d; j++)
        beta[j] = A[j][d];
    return beta;
}
vector<double> explain(function<double(vector<double>)> model, const vector<double> &x, int num_samples = 500, double sigma = 0.1)
{
    int d = x.size();
    vector<vector<double>> X;
    vector<double> y, w;
    mt19937 rng(123);
    normal_distribution<double> nd(0, sigma);
    for (int s = 0; s < num_samples; s++)
    {
        vector<double> z(d);
        for (int i = 0; i < d; i++)
            z[i] = x[i] + nd(rng);
        double p = model(z);
        vector<double> feat(d);
        for (int i = 0; i < d; i++)
            feat[i] = z[i] - x[i];
        X.push_back(feat);
        y.push_back(p);
        double dist = 0;
        for (int i = 0; i < d; i++)
        {
            double diff = z[i] - x[i];
            dist += diff * diff;
        }
        dist = sqrt(dist);
        w.push_back(exp(-dist * dist / (2 * sigma * sigma)));
    }
    return linear_regression(X, y, w);
}
int main()
{
    auto model = [](vector<double> z)
    { return z[0] + 2 * z[1] > 0 ? 1.0 : 0.0; };
    vector<double> x{0.1, -0.2};
    auto beta = explain(model, x);
    for (size_t i = 0; i < beta.size(); ++i)
    {
        if (i)
            cout << " ";
        cout << beta[i];
    }
    cout << "\n";
}

File: X-AI/LIME/Python/lime.py
import random,math

def dot(a,b):
    return sum(x*y for x,y in zip(a,b))

def linear_regression(X,y,w):
    d=len(X[0])
    A=[[0.0]*d for _ in range(d)]
    b=[0.0]*d
    for i in range(len(X)):
        wi=w[i]
        for j in range(d):
            b[j]+=wi*y[i]*X[i][j]
            for k in range(d):
                A[j][k]+=wi*X[i][j]*X[i][k]
    for j in range(d):
        A[j].append(b[j])
    for i in range(d):
        piv=i
        for r in range(i+1,d):
            if abs(A[r][i])>abs(A[piv][i]):
                piv=r
        A[i],A[piv]=A[piv],A[i]
        if abs(A[i][i])<1e-12:
            continue
        fac=A[i][i]
        for c in range(i,d+1):
            A[i][c]/=fac
        for r in range(d):
            if r==i:
                continue
            fac=A[r][i]
            for c in range(i,d+1):
                A[r][c]-=fac*A[i][c]
    return [A[j][d] for j in range(d)]

def explain(model,x,num_samples=500,sigma=0.1):
    d=len(x)
    X=[]
    y=[]
    w=[]
    for _ in range(num_samples):
        z=[xi+random.gauss(0,sigma) for xi in x]
        p=model(z)
        X.append([zj - xi for zj,xi in zip(z,x)])
        y.append(p)
        dist=math.sqrt(sum((zj-xi)*(zj-xi) for zj,xi in zip(z,x)))
        w.append(math.exp(-dist*dist/(2*sigma*sigma)))
    beta=linear_regression(X,y,w)
    return beta

if __name__=="__main__":
    def model(z):
        return 1.0 if z[0]+2*z[1]>0 else 0.0
    x=[0.1,-0.2]
    print(explain(model,x))